plugins/java-decompiler/engine/src/org/jetbrains/java/decompiler/modules/decompiler/exps/NewExprent.java

```diff
@@ -205,6 +205,7 @@ public TextBuffer toJava(int indent, BytecodeMappingTracer tracer) {
           }
         }
 
+        boolean forceSameType = DecompilerContext.getOption(IFernflowerPreferences.OVERLOADING_SAFETY_CASTS);
```

Comment:
For constructor invocations, overloads from a base class are impossible. So it should be possible to adopt the ambiguity check from InvocationExprent.